transport/src/main/java/io/netty/channel/SingleThreadEventLoop.java

```diff
@@ -89,11 +103,81 @@ public ChannelFuture register(final Channel channel, final ChannelPromise promis
         return promise;
     }
 
+    /**
+     * Adds a task to be run once at the end of next (or current) {@code eventloop} iteration.
+     *
+     * @param task to be added.
+     */
+    final void onEventLoopIteration(Runnable task) {
+        ObjectUtil.checkNotNull(task, "task");
+        if (isShutdown() || isShuttingDown() || isTerminated()) {
```

Comment:
this involves 3 volatile reads. Can we reduce to a single read and add some new methods to support this?

```java
// in SingleThreadEventExecutor
protected final boolean canAddTask() {
  int state = this.state;
  return !isShutdown0(state) && ! ...;
}

public boolean isShutdown() {
  return isShutdown0(state);
}

private boolean isShutdown0(int state) {
  return state >= ST_SHUTDOWN;
}

...

// in SingleThreadEventLoop
final void onEventLoopIteration(Runnable task) {
  if (!canAddTask()) {
    reject(task);
  }
...
}
```

also we are guarding against different conditions than `SingleThreadEventExecutor.offerTask` (it only checks isShutdown()) ... is this intentional?